### PR TITLE
Permit searches to be concurrent to prewarming

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today when prewarming a searchable snapshot we use the `SparseFileTracker` to
lock each (part of a) snapshotted blob, blocking any other readers from
accessing this data until the whole part is available.

This commit changes this strategy: instead we optimistically start to download
the blob without any locking, and then lock much smaller ranges after each
individual `read()` call. This may mean that some bytes are downloaded twice,
but reduces the time that other readers may need to wait before the data they
need is available.

As a best-effort optimisation we try to request the smallest possible single
range of missing bytes in the part by first checking how many of the initial
and terminal bytes of the part are already present in cache. In particular if
the part is already fully cached before prewarming then this check means we
skip the part entirely.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:227]: I don't think there's a problem with the very fine-grained locking done here, but if it turns out to be a bottleneck then we can instead lock larger chunks, maybe ~1MB or so, rather than locking for each individual call to `read()`. That makes it a bit more complicated since we must worry about aligning reads to the chunk size too, but it's still feasible.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:229]: Maybe remove  the "now" word. Technically the range might have been already available before this got called?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:227]: Can we capture the returned Future here and assert in the next line that `future.isDone`?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:254]: I'm confused by the stat accounting here as I thought that not all `totalBytesRead` have been written to filechannel (when other searches have filled the bytes e.g.)?

**Review by ywelsch — COMMENTED** [review]
Thanks David. I've left some minor comments, looking good o.w. I think that @tlrx should have a look too before merge though.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Sure, done.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:227]: No, the future might not be done on return if another thread already acquired the target range and hasn't filled it in yet. That's ok, tho, we assume it'll succeed eventually and keep on consuming the blob until we find a range that hasn't already been acquired by anyone else.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:254]: 😕 I thought the intention is for this stat to record the bytes we downloaded from the blob store and so I intended not to account solely for the bytes we actually wrote to the cache file. But I think it's not an accurate count of the bytes downloaded either. I can make it account only for the bytes written by the prewarmer. @tlrx WDYT?

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:254]: I think that `cached_bytes_written` should reflect the number of bytes effectively written in cache. It reflected the number of bytes downloaded (only when the cache range size and the `readBlobPreferredLength` have the same value) but I think we should compute the download stats differently.

**Review by tlrx — APPROVED** [review]
LGTM, thanks David.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java]: Ok, I adjusted this in cd793b3 and am checking that the tests are still solid now.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:229]: I will leave it as-is, we'll be doing some more general benchmarking that will tell us if it needs further optimisation.

**Pull request merged**
